src/test/java/org/cactoos/io/ZipTest.java

```diff
@@ -24,51 +24,87 @@
 
 /**
  * Test case for {@link Zip}.
  *
  * @since 0.29
  * @checkstyle JavadocMethodCheck (500 lines)
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
  */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
 public final class ZipTest {
+
     /**
      * Temporary folder.
      */
     @Rule
-    public final TemporaryFolder folder = new TemporaryFolder();
+    public final TemporaryFolder temporal = new TemporaryFolder();
 
     @Test
-    public void zip() throws Exception {
-        final Path dir = this.folder.newFolder().toPath();
-        dir.resolve("x/y").toFile().mkdirs();
-        Files.write(dir.resolve("x/y/test"), "".getBytes());
-        try (ZipInputStream input = new ZipInputStream(
-            new Zip(
-                new Directory(dir)
-            ).stream()
+    public void mustZipDirectory() throws Exception {
+        final String zipname = "abc.zip";
+        final File folder = this.temporal.newFolder("abc");
+        try (OutputStream out = new OutputStreamTo(new File(folder, "A.txt"))) {
+            out.write("ABC".getBytes());
+        }
+        new File(folder, "B").mkdirs();
+        try (OutputStream out = new OutputStreamTo(
+            new File(
+                folder,
+                new Joined(
+                    File.separator, "B", "B.txt"
+                ).asString()
+            )
         )) {
-            int cnt = 0;
-            ZipEntry entry = input.getNextEntry();
-            while (entry != null) {
-                ++cnt;
-                entry = input.getNextEntry();
-            }
-            MatcherAssert.assertThat(
-                "Can't list files in a directory represented by a path",
-                cnt,
-                // @checkstyle MagicNumber (1 line)
-                Matchers.equalTo(4)
-            );
+            out.write("EFG".getBytes());
         }
+        new File(folder, "C").mkdirs();
+        new Assertion<>(
+            "Must zip directory with the same directory structure",
+            new Sticky<>(
+                () -> {
+                    try (OutputStream out = new OutputStreamTo(
+                        this.temporal.newFile(zipname)
+                    )) {
+                        out.write(
+                            new BytesOf(
+                                new Zip(new Directory(folder))
+                            ).asBytes()
+                        );
+                    }
+                    try (ZipFile zipped = new ZipFile(
+                        new File(folder.getParentFile(), zipname)
+                    )
+                    ) {
+                        return new IterableOf<>(
+                            zipped.stream().map(
+                                e -> new TextOf(e.getName())
+                            ).collect(Collectors.toList())
+                        );
+                    }
+                }
+            ),
+            new ScalarHasValue<>(
```

Comment:
@fanifieiev can't we get rid of the `Scalar` stuffs in that test? It seems over-engineered

Reply:
@victornoel I tried, but this one was the most appropriate in this situation

Reply:
@fanifieiev ok, let's keep it like that, I don't see any way to make the test clearer except moving the zipping part out of the assertion argument.